service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6259,259 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
 
 	return RefreshTasksForWorkflowStart(ctx, ms, ms.taskGenerator, EmptyVersionedTransition)
 }
+
+func (ms *MutableStateImpl) currentVersionedTransition() *persistencespb.VersionedTransition {
+	if len(ms.executionInfo.TransitionHistory) == 0 {
+		// transition history is not enabled
+		return nil
+	}
+	transitionHistory := ms.executionInfo.TransitionHistory
+	return transitionHistory[len(transitionHistory)-1]
+}
+
+func (ms *MutableStateImpl) ApplyMutation(
+	mutation *persistencespb.WorkflowMutableStateMutation,
+) error {
+	prevExecutionInfoSize := ms.executionInfo.Size()
+	currentVersionedTransition := ms.currentVersionedTransition()
+
+	ms.applySignalRequestedIds(mutation.SignalRequestedIds, mutation.ExecutionInfo)
+	ms.applyTombstones(mutation.SubStateMachineTombstoneBatches, currentVersionedTransition)
+	ms.syncExecutionInfo(ms.executionInfo, mutation.ExecutionInfo)
```

Comment:
`SubStateMachineTombstoneBatches` is empty for mutation.ExecutionInfo as we have a separate `mutation.SubStateMachineTombstoneBatches`